### PR TITLE
Fixed parsing of 802.11 frames with Country Information element tags that did not add the padding byte

### LINKED ISSUE #4132: 802.11 Country Information Element - Wrong Handling of Missing Pads
### Brief description

When parsing a 802.11 beacon frame with a Country Information Tag, scapy expects a padding byte if the tag length is odd (see https://github.com/secdev/scapy/blob/master/scapy/layers/dot11.py#L1285). This is correct, per the standard.
Some transmitters don't comply with the standard and don't add the padding byte, causing scapy to fail, because it treats the next tag ID byte as the padding byte.

Wireshark parses those without failing - because instead of checking whether a padding byte should exist and relying on its existence (if required), it just checks whether there's padding: https://github.com/boundary/wireshark/blob/master/epan/dissectors/packet-ieee80211.c#L11208

### Scapy version

2.5.0

### Python version

3.8.9

### Operating system

Ubuntu

### Additional environment information

_No response_

### How to reproduce

Use scapy 2.5.0 to parse the attached file.
2. Expected: 
3. A

### Actual result

Bad parsing, a 2nd SSID tag is found.

### Expected result

Proper parsing all the way to the last information element tag.

### Related resources

[scapy_exception_beacon.pcapng.gz](https://github.com/secdev/scapy/files/12719039/scapy_exception_beacon.pcapng.gz)

### PR DESCRIPTION
**Pull request by giladbeeri**
Fixed handling of 802.11 frames with Country Information element tags that don't comply to the standard, by having an odd length, and not adding the padding byte.

fixes #4132

### CODE REVIEW

File: scapy/layers/dot11.py [hunk 1]
```diff
@@ -1282,7 +1310,7 @@ class Dot11EltCountry(Dot11Elt):
         # When this extension is last, padding appears to be omitted
         ConditionalField(
             ByteField("pad", 0),
-            lambda pkt: (len(pkt.descriptors) + 1) % 2
+            country_info_element_padding_condition
```
Comment (gpotter2) [inline scapy/layers/dot11.py:1313]: I'd do something simpler in one line
```
lambda pkt: ((len(pkt.descriptors) + 1) % 2) if pkt.len is None else (pkt.len % 3)
```
just to handle both dissecting and build cases. Wouldn't that work?

File: test/regression.uts [hunk 2]
```diff
@@ -1427,13 +1427,13 @@ conf.checkIPsrc = conf_checkIPsrc
 = Padding assembly
 r = raw(Padding("abc"))
 r
-assert r == b"abc"
```
Comment (gpotter2) [inline test/regression.uts:1430]: Please don't do unnecessary changes.

File: test/regression.uts [hunk 3]
```diff
@@ -1568,6 +1568,23 @@ beacon = frame.getlayer(5)
...
 
+= SSID is parsed properly even when the Country Information Tag Element has an odd length (not complying with the standard) and a missing pad byte
+~ Missing Pad Byte in Country Info
+# A regression test for https://github.com/secdev/scapy/pull/2685.
+# https://github.com/secdev/scapy/issues/4132 describes a packet with
+# a Country Information element tag that has an odd length, even though it's against the standard.
+# The transmitter should have added a padding byte to make the length even, but it didn't.
+# The effect on scapy used to be improper parsing of the next tag elements, causing the SSID to be overridden.
+# This test checks the SSID is parsed properly.
+filename = scapy_path("/test/pcaps/dot11_country_info_missing_pad_byte.pcap")
```
Comment (gpotter2) [inline test/regression.uts:1579]: For such a small packet there's no need to put a pcap, you should embed the packet in the test as bytes.

File: test/regression.uts [hunk 4]
```diff
@@ -2939,17 +2956,17 @@ def test_solaris_111(mock_os, mock_get_if_addr):
...
+-------------------- --------------- -------------------- ------ ----- --- --- ----- ------
+default              0.0.0.0         10.0.2.2             net0    1500   2 UG       5      0
+10.0.2.0             255.255.255.0   10.0.2.15            net0    1500   3 U        0      0
 127.0.0.1            255.255.255.255 127.0.0.1            lo0     8232   2 UH    1517   1517
 """
     # Mocked file descriptor
     strio = StringIO(netstat_output)
     mock_os.popen = mock.MagicMock(return_value=strio)
     print(scapy.arch.unix.SOLARIS)
-
```
Comment (gpotter2) [inline test/regression.uts:2952]: Please cancel all of those.

**Review by gpotter2 — CHANGES_REQUESTED** [review]

**Comment by gpotter2** [comment]
Thanks for the PR ! I left a few comments.

File: scapy/layers/dot11.py (hunk 1 above)
Comment (giladbeeri) [reply scapy/layers/dot11.py:1317]: I added the `is None` test for the build case.
I think the long version is more readable because it explains to the reader the logic behind the condition, and it also explicitly fails for scenarios such as `pkt.len == 83`. In my version, a `ValueError` is expected, but in your version, the dissector will assume there's a single padding byte, even though it is wrong (because if the length is 83 and there's a padding byte, then it is 82, but 82 isn't divided by 3). At the moment, we don't know if there's a case with % == 2 (e.g., 83), so your version also works.

Comment (gpotter2) [reply scapy/layers/dot11.py:1317]: Isn't the padding byte included in the length?

**Comment by gpotter2** [comment]
I still really dislike the long version.
Would you mind switching to a one liner, and adding a shorter version of your comment above it?

File: scapy/layers/dot11.py (hunk 1 above)
Comment (giladbeeri) [reply scapy/layers/dot11.py:1317]: The padding byte is indeed included, it can 83 if the transmitter is non-compliant. This PR started because we found "in the wild" a non-compliant transmitter, using a length of 81, not adding the required padding byte (by the length).
I'll change to a one-liner, but the cost is that handling of non-compliant transmitters is worse.

**Review by gpotter2 — APPROVED** [review]
Thanks for the update !
